src/dune_pkg/toolchain.ml

```diff
@@ -105,78 +104,157 @@ module Compiler = struct
-      make version [ "-j" ]
-    ;;
+  (* Run `make` in the compiler's source directory, passing it a given
+     list of arguments. *)
+  let make t args =
+    let make = Make.path ~path:env_path in
+    let source_dir = Path.outside_build_dir (Paths.source_dir t) in
+    let command = { Command.prog = `Absolute make; args } in
+    Command.run ~dir:source_dir command
+  ;;
 
-    (* Installation happens in two steps. First, run `make install
-       DESTDIR=...` to install the toolchain into a temporary directory. Then
-       the target directory from the temporary directory is copied into the
-       final installation directory. This allows us to use the fact that the
-       final installation directory exists to check that the toolchain is
-       installed. *)
-    let install version =
-      let open Fiber.O in
-      let dest_dir = Path.outside_build_dir (Paths.tmp_install_dir version) in
-      let target_dir = Path.outside_build_dir (Paths.target_dir version) in
-      Fpath.rm_rf (Path.to_string target_dir);
-      Fpath.rm_rf (Path.to_string dest_dir);
-      let+ () =
-        make version [ "install"; sprintf "DESTDIR=%s" (Path.to_string dest_dir) ]
-      in
-      Path.rename
-        (Path.outside_build_dir (Paths.target_dir_within_tmp_install_dir version))
-        (Path.outside_build_dir (Paths.target_dir version));
-      Fpath.rm_rf (Path.to_string dest_dir)
-    ;;
-  end
+  (* Install the compiler in the appropriate toolchain directory.
+
+     Installation happens in two steps. First, run `make install
+     DESTDIR=...` to install the toolchain into a temporary
+     directory. Then the target directory from the temporary directory
+     is copied into the final installation directory. This allows us
+     to use the fact that the final installation directory exists to
+     check that the toolchain is installed.
+
+     Note that the install commands from the compiler's opam file are
+     ignored here, and assumed to simply run `make install`. *)
+  let install t =
+    let open Fiber.O in
+    let dest_dir = Path.outside_build_dir (Paths.tmp_install_dir t) in
+    let target_dir = Path.outside_build_dir (Paths.target_dir t) in
+    Fpath.rm_rf (Path.to_string target_dir);
+    Fpath.rm_rf (Path.to_string dest_dir);
+    let+ () = make t [ "install"; sprintf "DESTDIR=%s" (Path.to_string dest_dir) ] in
+    Path.rename
+      (Path.outside_build_dir (Paths.target_dir_within_tmp_install_dir t))
+      (Path.outside_build_dir (Paths.target_dir t));
+    Fpath.rm_rf (Path.to_string dest_dir)
+  ;;
 
   let bin_dir t = Path.Outside_build_dir.relative (Paths.target_dir t) "bin"
   let is_version_installed t = Path.exists (Path.outside_build_dir (Paths.target_dir t))
 
-  let of_version_and_resolved_package version resolved_package =
+  let checksum { opam_file; _ } =
+    Option.bind opam_file.url ~f:(fun opam_file_url ->
+      OpamFile.URL.checksum opam_file_url
+      |> List.hd_opt
```

Comment:
Good point, but currently dune's `Fetch.fetch` only accepts a single checksum. I guess we could make sure we pass it the "best" checksum rather than just the first one. Eventually we should probably have a fetch function that checks all the checksums too.

Reply:
I updated this to use the strongest available hash rather than the first one in the file.

Reply:
I don't think picking the strongest is great either (also because what is strongest is a bit of a fluctuating scale, whether e.g. SHA3 is stronger than SHA2 is dependent on discoveries of issues and not a clear-cut criterium), and also because that allows an attacker to inject "stronger" variants of fake files that won't pass validation with "weaker" algorithms. In the long term we should do the same as OPAM and validate all supplied, but for now I guess this will do.